crates/rust-analyzer/src/diagnostics/to_proto.rs

```diff
@@ -61,21 +61,51 @@ fn location(
     let file_name = resolve_path(config, workspace_root, &span.file_name);
     let uri = url_from_abs_path(&file_name);
 
-    // FIXME: this doesn't handle UTF16 offsets correctly
     let range = lsp_types::Range::new(
-        lsp_types::Position::new(
-            (span.line_start as u32).saturating_sub(1),
-            (span.column_start as u32).saturating_sub(1),
-        ),
-        lsp_types::Position::new(
-            (span.line_end as u32).saturating_sub(1),
-            (span.column_end as u32).saturating_sub(1),
-        ),
+        position(config, span, span.line_start, span.column_start),
+        position(config, span, span.line_end, span.column_end),
     );
 
     lsp_types::Location { uri, range }
 }
 
+fn position(
+    config: &DiagnosticsMapConfig,
+    span: &DiagnosticSpan,
+    line_offset: usize,
+    column_offset: usize,
+) -> lsp_types::Position {
+    let line_index = line_offset - span.line_start;
+
+    let mut true_column_offset = column_offset;
+    if let Some(line) = span.text.get(line_index) {
+        if line.text.chars().count() == line.text.len() {
+            // all utf-8
+            return lsp_types::Position {
+                line: (line_offset as u32).saturating_sub(1),
+                character: (column_offset as u32).saturating_sub(1),
+            };
+        }
+        let mut char_offset = 0;
+        for char in line.text.chars() {
+            char_offset += 1;
+            if char_offset > column_offset {
+                break;
+            }
+            let offset_shift = match config.offset_encoding {
+                OffsetEncoding::Utf8 => char.len_utf8() - 1,
```

Comment:
I think we can skip the entire mapping logic in case of utf-8.

Reply:
sorry, i don't understand how to skip the entire mapping logic in case of utf-8. 
But we can put `match` before `for` to get the func pointer, therefore `match` will run only once.

Reply:
The old implementation of this function works fine for utf-8, right?

Reply:
yes

Reply:
I mean something like

```rust
fn position(
    config: &DiagnosticsMapConfig,
    span: &DiagnosticSpan,
    line_offset: usize,
    column_offset: usize,
) -> lsp_types::Position {
    if config.offset_encoding == OffsetEncoding::Utf8 {
        return lsp_types::Position::new(
            (span.line_start as u32).saturating_sub(1),
            (span.column_start as u32).saturating_sub(1),
        );
    }

    // ...
}
```

Reply:
source code may contain three bytes or four bytes utf-8 character, for this situation, we cannot use span.column_start.

Reply:
Why not? `column_start` would use the byte offset in case of `OffsetEncoding::Utf8`, right? This is exactly what rustc returns in it's json diagnostics.